### PR TITLE
Add type hinting for jinja filters

### PR DESCRIPTION
**Pull request by rmccar**
### What is the context of this PR?
This PR will add type hinting to jinja filters

### How to review 
- Changes make sense
- Run `run_mypy.sh` see that theres no errors

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

**Comment by rmccar** [comment]
[Add type hinting for jinja filters (2 SP)](https://trello.com/c/mMcfE5TO/5309-add-type-hinting-for-jinja-filters-2-sp)

File: app/jinja_filters.py [hunk 1]
```diff
@@ -1,41 +1,48 @@
...
 
 
-def get_formatted_address(address_fields):
+def get_formatted_address(address_fields: dict) -> str:
     address_fields.pop("uprn", None)
     return "<br>".join(address_field for address_field in address_fields.values())
 
 
-def get_formatted_currency(value, currency="GBP") -> str:
+def get_formatted_currency(value: Union[float, Decimal], currency: str = "GBP") -> Any:
```
Comment (berroar) [inline app/jinja_filters.py:45]: The return here should be str, might need explicitly set the var as a str for this to work though.

File: app/jinja_filters.py [hunk 2]
```diff
@@ -134,7 +149,7 @@ def get_format_date(value):
 
 @pass_eval_context  # type: ignore
 @blueprint.app_template_filter()
-def format_datetime(context, date_time):
+def format_datetime(context: str, date_time: str) -> str:
```
Comment (berroar) [inline app/jinja_filters.py:152]: I think date_time is type `datetime` and not `str` here

File: app/jinja_filters.py [hunk 3]
```diff
@@ -148,7 +163,7 @@ def format_datetime(context, date_time):
     return mark_safe(context, result)
 
 
-def get_format_date_range(start_date, end_date):
+def get_format_date_range(start_date: str, end_date: str) -> flask_babel:
```
Comment (berroar) [inline app/jinja_filters.py:166]: These appear to be of type `Markup`

File: app/jinja_filters.py [hunk 4]
```diff
@@ -157,35 +172,35 @@ def get_format_date_range(start_date, end_date):
...
-def setAttributes(dictionary, attributes):
+def setAttributes(dictionary: dict, attributes: dict) -> dict:
     for key in attributes:
         dictionary[key] = attributes[key]
     return dictionary
 
 
 @blueprint.app_template_filter()
-def should_wrap_with_fieldset(question):
+def should_wrap_with_fieldset(question: dict) -> Union[Any, bool]:
```
Comment (berroar) [inline app/jinja_filters.py:203]: I think this is just returning `bool`

**Review by berroar — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py [hunk 5]
```diff
@@ -1,41 +1,48 @@
...
 blueprint = flask.Blueprint("filters", __name__)
 
 
-def mark_safe(context, value):
+def mark_safe(context: Any, value: str) -> str:
     return Markup(value) if context.autoescape else value
 
 
-def strip_tags(value):
+def strip_tags(value: str) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:24]: This returns escaped Markup if you check escape() method return type what type does it return?

File: app/forms/validators.py [hunk 6]
```diff
@@ -445,7 +454,7 @@ def _is_valid(
 
 def format_playback_value(
     value: Union[float, Decimal], currency: Optional[str] = None
-) -> str:
+) -> Any:
```
Comment (berroar) [inline app/forms/validators.py:457]: This should return a str, as we are enforcing the var being returned as a string.

File: app/jinja_filters.py [hunk 7]
```diff
@@ -1,41 +1,48 @@
...
     return escape(Markup(value).striptags())
 
 
 @blueprint.app_template_filter()
-def format_number(value):
+def format_number(value: int) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:29]: This can accept both integers and floats as it uses `format_decimal()` further down which accepts floats as arguments.

File: app/jinja_filters.py [hunk 8]
```diff
@@ -45,48 +52,56 @@ def get_formatted_currency(value, currency="GBP") -> str:
...
+def get_currency_symbol(currency: str = "GBP") -> str:
+    currency_symbol: str = numbers.get_currency_symbol(
+        currency, locale=flask_babel.get_locale()
+    )
+    return currency_symbol
 
 
 @blueprint.app_template_filter()
-def format_percentage(value):
+def format_percentage(value: str) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:63]: The value here can be an `int` or `Decimal`, then it is converted to `str`

File: app/jinja_filters.py [hunk 9]
```diff
@@ -45,48 +52,56 @@ def get_formatted_currency(value, currency="GBP") -> str:
...
     return f"{value}%"
 
 
-def format_unit(unit, value, length="short"):
-    return units.format_unit(
+def format_unit(unit: str, value: str, length: str = "short") -> str:
```
Comment (petechd) [inline app/jinja_filters.py:67]: `value` here can be `Decimal` or `int`.

File: app/jinja_filters.py [hunk 10]
```diff
@@ -45,48 +52,56 @@ def get_formatted_currency(value, currency="GBP") -> str:
...
+            measurement_unit=unit,
+            length=unit_length,
+            locale=flask_babel.get_locale(),
+        ).strip()
+
+    return unit_label
 
 
-def format_duration(value):
+def format_duration(value: Mapping) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:104]: Needs at least the top-level type parameters.

**Review by petechd — CHANGES_REQUESTED** [review]

File: app/forms/validators.py (hunk 6 above)
Comment (rmccar) [reply app/forms/validators.py:457]: This needed to be any because the return on 459 returns `get_formatted_currency` which needs to return Any because it uses a function from babel that returns Any

File: app/jinja_filters.py [hunk 11]
```diff
@@ -236,19 +252,25 @@ def get_width_for_number(answer):
...
         self.description = description
 
 
 class SelectConfig:
-    def __init__(self, option, index, answer, form=None):
+    def __init__(
+        self,
+        option: SelectFieldBase._Option,
+        index: int,
+        answer: Mapping,
```
Comment (petechd) [inline app/jinja_filters.py:271]: These  mappings need top level parameters.

File: app/jinja_filters.py [hunk 12]
```diff
@@ -295,7 +319,11 @@ def __init__(self, option, index, answer):
 
 
 class OtherConfig:
-    def __init__(self, detail_answer_field, detail_answer_schema):
+    def __init__(
+        self,
+        detail_answer_field: SelectFieldBase._Option,
+        detail_answer_schema: Mapping,
```
Comment (petechd) [inline app/jinja_filters.py:325]: Mapping needs top level parameters (as described in type hinting doc in runner).

File: app/jinja_filters.py [hunk 13]
```diff
@@ -320,7 +348,7 @@ def __init__(self, detail_answer_field, detail_answer_schema):
 
 
 @blueprint.app_template_filter()  # type: ignore
-def map_select_config(form, answer):
+def map_select_config(form: Mapping, answer: Mapping) -> list[SelectConfig]:
```
Comment (petechd) [inline app/jinja_filters.py:351]: Same here, mappings need top level parameters.

File: app/jinja_filters.py [hunk 14]
```diff
@@ -330,12 +358,14 @@ def map_select_config(form, answer):
...
 @blueprint.app_context_processor
-def map_select_config_processor():
+def map_select_config_processor() -> dict:
     return dict(map_select_config=map_select_config)
 
 
 @blueprint.app_template_filter()  # type: ignore
-def map_relationships_config(form, answer):
+def map_relationships_config(
+    form: Mapping, answer: Mapping
```
Comment (petechd) [inline app/jinja_filters.py:367]: Mappings need top level parameters.

**Review by petechd — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py [hunk 15]
```diff
@@ -534,27 +573,25 @@ def map_summary_item_config(
...
-    list_items,
-    icon,
-    edit_link_text=None,
-    edit_link_aria_label=None,
-    remove_link_text=None,
-    remove_link_aria_label=None,
-):
+    list_items: list,
+    icon: str,
+    edit_link_text: Optional[str] = "",
```
Comment (petechd) [inline app/jinja_filters.py:590]: These two `Optional[str]` have default value of `""` so they shouldn't be optional. Also `list_items: list` needs top level parameter.

File: app/jinja_filters.py [hunk 16]
```diff
@@ -534,27 +573,25 @@ def map_summary_item_config(
...
+    edit_link_aria_label: str = "",
+    remove_link_text: Optional[str] = "",
+    remove_link_aria_label: str = "",
+) -> list:
```
Comment (petechd) [inline app/jinja_filters.py:594]: Returned `list` needs type of the top level parameter as well.

File: app/jinja_filters.py [hunk 17]
```diff
@@ -603,5 +640,5 @@ def map_list_collector_config(
 
 
 @blueprint.app_context_processor
-def map_list_collector_config_processor():
+def map_list_collector_config_processor() -> dict:
```
Comment (petechd) [inline app/jinja_filters.py:643]: Again, we need to know what this generic `dict` returns.

File: app/jinja_filters.py [hunk 18]
```diff
@@ -534,27 +573,25 @@ def map_summary_item_config(
...
-            SummaryRow(calculated_question, summary_type, False, None, None, None)
-        )
+        rows.append(SummaryRow(calculated_question, summary_type, False, "", "", ""))
 
     return rows
 
 
 @blueprint.app_context_processor
-def map_summary_item_config_processor():
+def map_summary_item_config_processor() -> dict:
```
Comment (petechd) [inline app/jinja_filters.py:582]: We need to know what generic type `dict` returns.

File: app/jinja_filters.py (hunk 16 above)
Comment (petechd) [reply app/jinja_filters.py:603]: This one is still outstanding.

File: app/jinja_filters.py (hunk 5 above)
Comment (petechd) [reply app/jinja_filters.py:26]: This one is still outstanding, it returns `Markup`.
Screenshot 2022-06-20 at 15 13 05

File: app/jinja_filters.py [hunk 19]
```diff
@@ -1,92 +1,110 @@
...
 
 @blueprint.app_template_filter()
-def format_percentage(value):
+def format_percentage(value: Union[int, Decimal]) -> str:
     return f"{value}%"
 
 
-def format_unit(unit, value, length="short"):
-    return units.format_unit(
+def format_unit(unit: Union[int, Decimal], value: str, length: str = "short") -> str:
```
Comment (petechd) [inline app/jinja_filters.py:70]: Here `value` can be integer but potentially also `Decimal`, not a string.

File: app/jinja_filters.py [hunk 20]
```diff
@@ -104,14 +122,14 @@ def format_duration(value):
...
-def get_format_multilined_string(value):
+def get_format_multilined_string(value: str) -> str:
     escaped_value = escape(value)
     new_line_regex = r"(?:\r\n|\r|\n)+"
     value_with_line_break_tag = re.sub(new_line_regex, "<br>", escaped_value)
     return f"{value_with_line_break_tag}"
 
 
-def get_format_date(value):
+def get_format_date(value: str) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:134]: `value` is `Markup` here.

File: app/jinja_filters.py [hunk 21]
```diff
@@ -134,7 +152,7 @@ def get_format_date(value):
 
 @pass_eval_context  # type: ignore
 @blueprint.app_template_filter()
-def format_datetime(context, date_time):
+def format_datetime(context: nodes.EvalContext, date_time: str) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:157]: Worth double checking but I think `date_time` parameter needs to be `datetime` type.

File: app/jinja_filters.py [hunk 22]
```diff
@@ -134,7 +155,7 @@ def get_format_date(value):
 
 @pass_eval_context  # type: ignore
 @blueprint.app_template_filter()
-def format_datetime(context, date_time):
+def format_datetime(context: nodes.EvalContext, date_time: datetime) -> str:
```
Comment (petechd) [inline app/jinja_filters.py:158]: This should have the same return value as the `markup_safe` method because that is what it returns.

File: app/jinja_filters.py [hunk 23]
```diff
@@ -148,7 +169,7 @@ def format_datetime(context, date_time):
     return mark_safe(context, result)
 
 
-def get_format_date_range(start_date, end_date):
+def get_format_date_range(start_date: Markup, end_date: Markup) -> flask_babel:
```
Comment (petechd) [inline app/jinja_filters.py:172]: It can't just have the whole `flask_babel` library as the return type.

File: app/jinja_filters.py [hunk 24]
```diff
@@ -212,12 +237,12 @@ def should_wrap_with_fieldset(question):
...
 
 @blueprint.app_context_processor
-def should_wrap_with_fieldset_processor():
+def should_wrap_with_fieldset_processor() -> dict[str, Callable]:
     return {"should_wrap_with_fieldset": should_wrap_with_fieldset}
 
 
 @blueprint.app_template_filter()
-def get_width_for_number(answer):
+def get_width_for_number(answer: AnswerType) -> int:
```
Comment (petechd) [inline app/jinja_filters.py:245]: This method not always has a return (if width is > allowable_width) so needs to be `Optional`.

File: app/jinja_filters.py (hunk 23 above)
Comment (petechd) [reply app/jinja_filters.py]: Does this one not return `gettext` so `flask_babel.gettext`?

Comment (rmccar) [reply app/jinja_filters.py]: I checked what was returned where this function was being called and it was some markup so I made this function return markup

Comment (petechd) [reply app/jinja_filters.py]: In my check it returns a string so should be a string or Union of string and Markup?
Screenshot 2022-06-21 at 14 47 04

File: schemas/test/en/test_date_validation_range.json [hunk 25]
```diff
@@ -47,7 +47,7 @@
                                 "guidance": {
                                     "contents": [
                                         {
-                                            "list": ["Enter a date range between 23 days and 1 month and 20 days apart"]
+                                            "description": "Enter a date range between 23 days and 1 month and 20 days apart"
```
Comment (petechd) [inline schemas/test/en/test_date_validation_range.json:50]: Unnecessary change committed?

Comment (rmccar) [reply schemas/test/en/test_date_validation_range.json:50]: No I noticed that in one of the test schemas the guidance was wrongly formatted so fixed it

File: app/jinja_filters.py [hunk 26]
```diff
@@ -534,27 +589,25 @@ def map_summary_item_config(
...
-    list_items,
-    icon,
-    edit_link_text=None,
-    edit_link_aria_label=None,
-    remove_link_text=None,
-    remove_link_aria_label=None,
-):
+    list_items: list[dict[str, Union[str, int]]],
+    icon: str,
+    edit_link_text: str = "",
```
Comment (berroar) [inline app/jinja_filters.py:606]: I think theses should probably be `Optional[str] = None` but will need to find a way to the `.format` method further down.

Comment (rmccar) [reply app/jinja_filters.py]: Yeah thats why I didn't make them `Optional[str] = None`.
Ive come up with a solution though and pushed it up

**Review by petechd — APPROVED** [review]

**Review by berroar — APPROVED** [review]
